Changes for Coq 8.13 (#13527)
coq/coq: doc/sphinx/changes.rst

   .. include:: ../unreleased.rst

Version 8.13
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Coq version 8.13 integrates many usability improvements, as well
as extensions of the core language.
The main changes include:

  - Introduction of :ref:`primitive persistent arrays<primitive-arrays>`
    in the core language, implemented using imperative persistent arrays.
  - Introduction of definitional proof irrelevance for the equality type
    defined in the SProp sort, using the :flag:`Definitional UIP` flag,
    (deactivated by default). This models definitional uniqueness of
    identity proofs for this type. It is deactivated by default as
    it results in non-termination in combination with impredicativity.

>     Use of this flag is also shown by :cmd:`Print Assumptions`.

Reviewer: ```suggestion
    This flag is also used by :cmd:`Print Assumptions`.
```

Not sure if that's what you meant.  The "shown" doesn't read well.

Author: printed is better I think, fixed.

>   - Cumulative record and inductives type declarations can now

Reviewer: ```suggestion
  - Cumulative record and inductive type declarations can now
```

Author: Fixed

    :ref:`specify<813VarianceDecl>` the  variance of their universes.
  - Various bugfixes and uniformization of behavior with respect to the use of
    implicit arguments and the handling of existential variables in
    declarations, unification and tactics.
  - New warning for :ref:`unused variables <813UnusedVar>` in catch-all match branches that match
    multiple distinct patterns.
  - New warning for `Hint` commands outside sections without a locality
    attribute, whose goal is to eventually remove the fragile default

>     behavior of having hints be imported when using `Require` only.

Reviewer: ```suggestion
    behavior of importing hints only when using `Require`.
```

at https://coq.github.io/doc/v8.13/api.

Maxime Dénès, Emilio Jesús Gallego Arias, Gaëtan Gilbert, Michael
Soegtrop and Théo Zimmermann worked on maintaining and improving the
continuous integration system and package building infrastructure.

Erik Martin-Dorel has maintained the `Coq Docker images
<https://hub.docker.com/r/coqorg/coq>`_ that are used in many Coq
projects for continuous integration.

The OPAM repository for Coq packages has been maintained by
Guillaume Claret, Karl Palmskog, Matthieu Sozeau and Enrico Tassi with
contributions from many users. A list of packages is available at
https://coq.inria.fr/opam/www/.

TODO updated maintainers
Our current 31 maintainers are Yves Bertot, Frédéric Besson, Tej
Chajed, Cyril Cohen, Pierre Corbineau, Pierre Courtieu, Maxime Dénès,
Jim Fehrle, Julien Forest, Emilio Jesús Gallego Arias, Gaëtan Gilbert,
Georges Gonthier, Benjamin Grégoire, Jason Gross, Hugo Herbelin,
Vincent Laporte, Assia Mahboubi, Kenji Maillard, Guillaume Melquiond,
Pierre-Marie Pédrot, Clément Pit-Claudel, Kazuhiko Sakaguchi, Vincent
Semeria, Michael Soegtrop, Arnaud Spiwack, Matthieu Sozeau, Enrico

> Tassi, Laurent Théry, Anton Trunov, Li-yao Xia, Théo Zimmermann

Reviewer: ```suggestion
Tassi, Laurent Théry, Anton Trunov, Li-yao Xia and Théo Zimmermann
```

Reviewer: You missed this one.

The 52 contributors to this version are Reynald Affeldt, Tanaka Akira, Frédéric
Besson, Lasse Blaauwbroek, Clément Blaudeau, Martin Bodin, Ali Caglayan, Tej Chajed,
Cyril Cohen, Julien Coolen, Matthew Dempsky, Maxime Dénès, Andres Erbsen,
Jim Fehrle, Emilio Jesús Gallego Arias, Paolo G. Giarrusso, Attila Gáspár, Gaëtan Gilbert,
Jason Gross, Benjamin Grégoire, Hugo Herbelin, Wolf Honore, Jasper Hugunin, Ignat Insarov,
Ralf Jung, Fabian Kunze, Vincent Laporte, Olivier Laurent, Larry D. Lee Jr,
Thomas Letan, Yishuai Li, Xia Li-yao, James Lottes, Jean-Christophe Léchenet,
Kenji Maillard, Erik Martin-Dorel, Yusuke Matsushita, Guillaume Melquiond,
Carl Patenaude-Poulin, Clément Pit-Claudel, Pierre-Marie Pédrot, Pierre Roux,
Kazuhiko Sakaguchi, Vincent Semeria, Michael Soegtrop, Matthieu Sozeau,

> Enrico Tassi, Anton Trunov, Edward Wang, Li-yao Xia, Beta Ziliani, Théo Zimmermann.

Reviewer: ```suggestion
Enrico Tassi, Anton Trunov, Edward Wang, Li-yao Xia, Beta Ziliani and Théo Zimmermann.
```

The Coq community at large helped improve the design of this new version via
the GitHub issue and pull request system, the Coq development mailing list
coqdev@inria.fr, the coq-club@inria.fr mailing list, the `Discourse forum
<https://coq.discourse.group/>`_ and the `Coq Zulip chat <http://coq.zulipchat.com>`_.

Version 8.13's development spanned 5 months from the release of
Coq 8.12.0. Enrico Tassi and Maxime Dénès are the release managers of Coq 8.13.

> This release is the result of 400 PRs merged, closing ~100 issues.

Reviewer: ```suggestion
This release is the result of 400 merged PRs, closing ~100 issues.
```

Author: Thanks @jfehrle for the proofreading, all fixed now

  - Improvements to the reference manual including updated syntax descriptions that
    match Coq's grammar in several chapters.

See the `Changes in 8.13`_ section and following sections for the
detailed list of changes, including potentially breaking changes marked
with **Changed**.

Coq's documentation is available at https://coq.github.io/doc/v8.13/refman (reference
manual), and https://coq.github.io/doc/v8.13/stdlib (documentation of
the standard library). Developer documentation of the ML API is available
at https://coq.github.io/doc/v8.13/api.

Maxime Dénès, Emilio Jesús Gallego Arias, Gaëtan Gilbert, Michael
Soegtrop and Théo Zimmermann worked on maintaining and improving the
continuous integration system and package building infrastructure.

Erik Martin-Dorel has maintained the `Coq Docker images
<https://hub.docker.com/r/coqorg/coq>`_ that are used in many Coq
projects for continuous integration.

The OPAM repository for Coq packages has been maintained by
Guillaume Claret, Karl Palmskog, Matthieu Sozeau and Enrico Tassi with
contributions from many users. A list of packages is available at
https://coq.inria.fr/opam/www/.

TODO updated maintainers

> Our current 31 maintainers are Yves Bertot, Frédéric Besson, Tej

Reviewer 2: Now 32 maintainers. Our new maintainer is Olivier Laurent (@olaure01). The rest of the list is unchanged.

Chajed, Cyril Cohen, Pierre Corbineau, Pierre Courtieu, Maxime Dénès,
Jim Fehrle, Julien Forest, Emilio Jesús Gallego Arias, Gaëtan Gilbert,
Georges Gonthier, Benjamin Grégoire, Jason Gross, Hugo Herbelin,
Vincent Laporte, Assia Mahboubi, Kenji Maillard, Guillaume Melquiond,
Pierre-Marie Pédrot, Clément Pit-Claudel, Kazuhiko Sakaguchi, Vincent
Semeria, Michael Soegtrop, Arnaud Spiwack, Matthieu Sozeau, Enrico

> Tassi, Laurent Théry, Anton Trunov, Li-yao Xia, Théo Zimmermann

Reviewer: ```suggestion
Tassi, Laurent Théry, Anton Trunov, Li-yao Xia and Théo Zimmermann
```

Jason Gross, Benjamin Grégoire, Hugo Herbelin, Wolf Honore, Jasper Hugunin, Ignat Insarov,
Ralf Jung, Fabian Kunze, Vincent Laporte, Olivier Laurent, Larry D. Lee Jr,
Thomas Letan, Yishuai Li, Xia Li-yao, James Lottes, Jean-Christophe Léchenet,
Kenji Maillard, Erik Martin-Dorel, Yusuke Matsushita, Guillaume Melquiond,
Carl Patenaude-Poulin, Clément Pit-Claudel, Pierre-Marie Pédrot, Pierre Roux,
Kazuhiko Sakaguchi, Vincent Semeria, Michael Soegtrop, Matthieu Sozeau,
Enrico Tassi, Anton Trunov, Edward Wang, Li-yao Xia, Beta Ziliani and Théo Zimmermann.

The Coq community at large helped improve the design of this new version via
the GitHub issue and pull request system, the Coq development mailing list
coqdev@inria.fr, the coq-club@inria.fr mailing list, the `Discourse forum
<https://coq.discourse.group/>`_ and the `Coq Zulip chat <http://coq.zulipchat.com>`_.

Version 8.13's development spanned 5 months from the release of
Coq 8.12.0. Enrico Tassi and Maxime Dénès are the release managers of Coq 8.13.
This release is the result of 400 merged PRs, closing ~100 issues.

| Nantes, November 2020,
| Matthieu Sozeau for the Coq development team
|

> Changes in 8.13
> ~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Changes in 8.13+beta1
~~~~~~~~~~~~~~~~~~~~~
```

.. contents::
   :local:

Kernel
^^^^^^

- **Added:**
  Definitional UIP, only when :flag:`Definitional UIP` is enabled. See
  documentation of the flag for details.
  (`#10390 <https://github.com/coq/coq/pull/10390>`_,
  by Gaëtan Gilbert).
- **Added:**
  Built-in support for persistent arrays, which expose a functional
  interface but are implemented using an imperative data structure, for
  better performance.
  (`#11604 <https://github.com/coq/coq/pull/11604>`_,
  by Maxime Dénès and Benjamin Grégoire, with help from Gaëtan Gilbert).
- **Fixed:**
  A loss of definitional equality for declarations obtained through
  :cmd:`Include` when entering the scope of a :cmd:`Module` or
  :cmd:`Module Type` was causing :cmd:`Search` not to see the included
  declarations
  (`#12537 <https://github.com/coq/coq/pull/12537>`_, fixes `#12525
  <https://github.com/coq/coq/pull/12525>`_ and `#12647
  <https://github.com/coq/coq/pull/12647>`_, by Hugo Herbelin).
- **Changed:** Primitive arrays are now irrelevant in their single
  polymorphic universe (same as a polymorphic cumulative list
  inductive would be) (`#13356
  <https://github.com/coq/coq/pull/13356>`_, fixes `#13354

>   <https://github.com/coq/coq/issues/13354>`_, by Gaëtan Gilbert).

Reviewer 2: I think that this changelog entry should be merged with the one above on support for primitive arrays. In any case, the use of **Changed** in this case is made inappropriate by the fact that it doesn't change a feature that existed in the previous version.

Author: Right

Specification language, type inference
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **Changed:**
  In :tacn:`refine`, new existential variables unified with existing ones are no
  longer considered as fresh. The behavior of :tacn:`simple refine` no longer depends on
  the orientation of evar-evar unification problems, and new existential variables
  are always turned into (unshelved) goals. This can break compatibility in
  some cases (`#7825 <https://github.com/coq/coq/pull/7825>`_, by Matthieu
  Sozeau, with help from Maxime Dénès, review by Pierre-Marie Pédrot and
  Enrico Tassi, fixes `#4095 <https://github.com/coq/coq/issues/4095>`_ and

>   `#4413 <https://github.com/coq/coq/issues/4413>`_).

Reviewer 2: Does this break the compatibility beyond the tactics `refine` and `simple refine` that are mentioned here? Otherwise, I think this should be moved to the "Tactics" section.

Author: You are right, will move it.

  by Jasper Hugunin).

  .. _813UnusedVar:

- **Added:**
  Warning on unused variables in pattern-matching branches of
  :n:`match` serving as catch-all branches for at least two distinct
  patterns.
  (`#12768 <https://github.com/coq/coq/pull/12768>`_,
  fixes `#12762 <https://github.com/coq/coq/issues/12762>`_,
  by Hugo Herbelin).
- **Removed:**
  Undocumented and experimental forward class hint feature ``:>>``.
  Use ``:>`` (see :n:`@of_type`) instead
  (`#13106 <https://github.com/coq/coq/pull/13106>`_,
  by Pierre-Marie Pédrot).
- **Fixed:**
  Implicit arguments taken into account in defined fields of a record type declaration
  (`#13166 <https://github.com/coq/coq/pull/13166>`_,
  fixes `#13165 <https://github.com/coq/coq/issues/13165>`_,
  by Hugo Herbelin).
- **Added:**
  Definition and (Co)Fixpoint now support the :attr:`using` attribute.
  It has the same effect as :cmd:`Proof using`, which is only available in
  interactive mode.
  (`#13183 <https://github.com/coq/coq/pull/13183>`_,
  by Enrico Tassi).
- **Removed:** The type given to :cmd:`Instance` is no longer automatically
  generalized over unbound and :ref:`generalizable <implicit-generalization>` variables.

>   Use :n:`Instance : \`{@type}` instead of :n:`Instance : @type` to get the old behaviour, or

Reviewer 2: This attempt to escape a backtick leads to the backslash being kept in the rendered HTML:

![2020-12-02-125240_703x98_scrot](https://user-images.githubusercontent.com/1108325/100869402-4bbb2d80-349d-11eb-81a9-6db96344864e.png)

It's weird, because I didn't find any other documented way of escaping backticks.
@cpitclaudel Can you help?

Author: I fixed this using verbatim text, it works :)

  enable the compatibility flag :flag:`Instance Generalized Output`.
  (`#13188 <https://github.com/coq/coq/pull/13188>`_, fixes `#6042

>   <https://github.com/coq/coq/issues/6042>`_, by Gaëtan Gilbert).

Author: Right

  by Hugo Herbelin).
- **Deprecated**
  ``Numeral.v`` is deprecated, please use ``Number.v`` instead.
  (`#12218 <https://github.com/coq/coq/pull/12218>`_,
  by Pierre Roux).
- **Changed**
  Rational and real constants are parsed differently.
  The exponent is now encoded separately from the fractional part
  using ``Z.pow_pos``. This way, parsing large exponents can no longer
  blow up and constants are printed in a form closer to the one in which they
  were parsed (i.e., ``102e-2`` is reprinted as such and not ``1.02``).
  (`#12218 <https://github.com/coq/coq/pull/12218>`_,
  by Pierre Roux).
- **Removed**
  OCaml parser and printer for real constants have been removed.
  Real constants are now handled with proven Coq code.
  (`#12218 <https://github.com/coq/coq/pull/12218>`_,
  by Pierre Roux).
- **Added:**
  :ref:`Number Notation <number-notations>` and :ref:`String Notation
  <string-notations>` commands now
  support parameterized inductive and non inductive types
  (`#12218 <https://github.com/coq/coq/pull/12218>`_,
  fixes `#12035 <https://github.com/coq/coq/issues/12035>`_,
  by Pierre Roux, review by Jason Gross and Jim Fehrle for the
  reference manual).

> - **Changed:**
>   Scope information is propagated in indirect applications to a
>   reference prefixed with :g:`@@`; this covers for instance the case
>   :g:`r.(@@p) t` where scope information from :g:`p` is now taken into
>   account for interpreting :g:`t` (`#12685
>   <https://github.com/coq/coq/pull/12685>`_, by Hugo Herbelin).

Reviewer 2: It looks like these double `@` were an attempt to escape the `@`, but the ambiguity of `@` only occurs in `:n:` contexts, so these should be made single `@`:
```suggestion
- **Changed:**
  Scope information is propagated in indirect applications to a
  reference prefixed with :g:`@`; this covers for instance the case
  :g:`r.(@p) t` where scope information from :g:`p` is now taken into
  account for interpreting :g:`t` (`#12685
  <https://github.com/coq/coq/pull/12685>`_, by Hugo Herbelin).
```

  by Hugo Herbelin).
- **Changed:**
  New model for ``only parsing`` and ``only printing`` notations with
  support for at most one parsing-and-printing or only-parsing
  notation per notation and scope, but an arbitrary number of
  only-printing notations
  (`#12950 <https://github.com/coq/coq/pull/12950>`_,
  fixes `#4738 <https://github.com/coq/coq/issues/4738>`_
  and `#9682 <https://github.com/coq/coq/issues/9682>`_
  and part 2 of `#12908 <https://github.com/coq/coq/issues/12908>`_,
  by Hugo Herbelin).
- **Fixed:**
  Issues in the presence of notations recursively referring to another
  applicative notations, such as missing scope propagation, or failure
  to use a notation for printing
  (`#12960 <https://github.com/coq/coq/pull/12960>`_,
  fixes `#9403 <https://github.com/coq/coq/issues/9403>`_
  and `#10803 <https://github.com/coq/coq/issues/10803>`_,
  by Hugo Herbelin).
- **Fixed:**
  Capture the names of global references by
  binders in the presence of notations for binders
  (`#12965 <https://github.com/coq/coq/pull/12965>`_,
  fixes `#9569 <https://github.com/coq/coq/issues/9569>`_,
  by Hugo Herbelin).
- **Deprecated:**

>   :n:`Numeral Notation`, please use :ref:`Number Notation <number-notations>` instead.

Reviewer 2: ```suggestion
  `Numeral Notation`, please use :cmd:`Number Notation` instead
```

  to notations which match a larger part of the term to abbreviate
  (`#12986 <https://github.com/coq/coq/pull/12986>`_,
  by Hugo Herbelin).
- **Fixed:**
  Preventing notations for constructors to involve binders
  (`#13092 <https://github.com/coq/coq/pull/13092>`_,
  fixes `#13078 <https://github.com/coq/coq/issues/13078>`_,
  by Hugo Herbelin).
- **Added:**
  The :n:`@binder` entry of :cmd:`Notation` can now be used in
  notations expecting a single (non-recursive) binder
  (`#13265 <https://github.com/coq/coq/pull/13265>`_,
  by Hugo Herbelin, see section :ref:`notations-and-binders` of the
  reference manual).
- **Fixed:** Notations understand universe names without getting
  confused by different imported modules between declaration and use
  locations (`#13415 <https://github.com/coq/coq/pull/13415>`_, fixes
  `#13303 <https://github.com/coq/coq/issues/13303>`_, by Gaëtan
  Gilbert).

Tactics
^^^^^^^

- **Added:**
  :tacn:`lia` is extended to deal with boolean operators e.g. `andb` or `Z.leb`.
  (As `lia` gets more powerful, this may break proof scripts relying on `lia` failure.)
  (`#11906 <https://github.com/coq/coq/pull/11906>`_,  by Frédéric Besson).
- **Added:**

>   :tacn`apply … in` supports several hypotheses

Reviewer 2: ```suggestion
  :tacn:`apply … in` supports several hypotheses
```

  (`#12246 <https://github.com/coq/coq/pull/12246>`_,
  by Hugo Herbelin; grants
  `#9816 <https://github.com/coq/coq/pull/9816>`_).
- **Removed:**

>   The deprecated and undocumented "prolog" tactic was removed

Reviewer 2: ```suggestion
  The deprecated and undocumented `prolog` tactic was removed
```

  (`#12399 <https://github.com/coq/coq/pull/12399>`_,
  by Pierre-Marie Pédrot).

> - **Removed:** Removed `info` tactic that was deprecated in 8.5.

Reviewer 2: ```suggestion
- **Removed:** `info` tactic that was deprecated in 8.5.
```

  (`#12423 <https://github.com/coq/coq/pull/12423>`_, by Jim Fehrle).
- **Added:**
  The :tacn:`zify` tactic can now be extended by redefining the `zify_pre_hook`
  tactic. (`#12552 <https://github.com/coq/coq/pull/12552>`_,
  by Kazuhiko Sakaguchi).
- **Added:**
  The :tacn:`zify` tactic provides support for primitive integers (module :g:`ZifyInt63`).
  (`#12648 <https://github.com/coq/coq/pull/12648>`_,  by Frédéric Besson).
- **Changed:**
  Giving an empty list of occurrences after :n:`in` in tactics is no
  longer permitted. Omitting the :n:`in` gives the same behavior
  (`#13237 <https://github.com/coq/coq/pull/13236>`_,
  fixes `#13235 <https://github.com/coq/coq/issues/13235>`_,
  by Hugo Herbelin).
- **Fixed:**

>   Avoiding exposing an internal name of the form :n:`_tmp` when applying the

Reviewer 2: ```suggestion
  Avoid exposing an internal name of the form :n:`_tmp` when applying the
```

  list (for set complement) is still supported.  Ex: "at -1 -2"
  is no longer supported but "at -1 2" is.
  (`#13403 <https://github.com/coq/coq/pull/13403>`_,
  by Jim Fehrle).
- **Removed:**
  A number of tactics that formerly accepted negative
  numbers as parameters now give syntax errors for negative
  values.  These include {e}constructor, do, timeout,
  9 {e}auto tactics and psatz*.
  (`#13417 <https://github.com/coq/coq/pull/13417>`_,
  by Jim Fehrle).

Tactic language
^^^^^^^^^^^^^^^

- **Fixed:**
  Printing of the quotation qualifiers when printing :g:`Ltac` functions
  (`#13028 <https://github.com/coq/coq/pull/13028>`_,
  fixes `#9716 <https://github.com/coq/coq/issues/9716>`_
  and `#13004 <https://github.com/coq/coq/issues/13004>`_,
  by Hugo Herbelin).
- **Added:**
  An if-then-else syntax to Ltac2
  (`#13232 <https://github.com/coq/coq/pull/13232>`_,
  fixes `#10110 <https://github.com/coq/coq/issues/10110>`_,
  by Pierre-Marie Pédrot).

SSReflect
^^^^^^^^^

- **Added:**
  SSReflect intro pattern ltac views ``/[dup]``, ``/[swap]`` and ``/[apply]``
  (`#13317 <https://github.com/coq/coq/pull/13317>`_,
  by Cyril Cohen).
- **Fixed:**

>   Working around a bug of interaction between + and /(ltac:(...)) cf #13458

Reviewer 2: ```suggestion
  Working around a bug of interaction between + and /(ltac:(...)) cf
  `#13458 <https://github.com/coq/coq/issues/13458>`_
```

  evars are not abstracted implicitly anymore and instead
  raise an error
  (`#13139 <https://github.com/coq/coq/pull/13139>`_,
  by Pierre-Marie Pédrot).
- **Added:**
  Added support for automatic insertion of coercions in :cmd:`Search`
  patterns. Additionally, head patterns are now automatically
  interpreted as types
  (`#13255 <https://github.com/coq/coq/pull/13255>`_,
  fixes `#13244 <https://github.com/coq/coq/issues/13244>`_,
  by Hugo Herbelin).
- **Added:**
  The :cmd:`Proof using` command can now be used without loading the
  Ltac plugin (`-noinit` mode)
  (`#13339 <https://github.com/coq/coq/pull/13339>`_,
  by Théo Zimmermann).
- **Added:**
  Clarify in the documentation that :cmd:`Add ML Path` is not exported to compiled files
  (`#13345 <https://github.com/coq/coq/pull/13345>`_,
  fixes `#13344 <https://github.com/coq/coq/issues/13344>`_,
  by Hugo Herbelin).
- **Changed:**
  Option `-native-compiler` of the configure script now impacts the
  default value of the `-native-compiler` option of coqc. The
  `-native-compiler` option of the configure script is added as an on demand
  value, which becomes the default, thus preserving the previous default
  behavior.

>   The stdlib is still precompiled when configuring with `-native-compiler`
>   yes. It is not precompiled otherwise.

Reviewer: ```suggestion
  The stdlib is still precompiled when configuring with `-native-compiler
  yes`. It is not precompiled otherwise.
```

   .. include:: ../unreleased.rst

Version 8.13
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Coq version 8.13 integrates many usability improvements, as well
as extensions of the core language.
The main changes include:

  - :ref:`Introduction <813PrimArrays>` of :ref:`primitive persistent arrays<primitive-arrays>`
    in the core language, implemented using imperative persistent arrays.
  - Introduction of :ref:`definitional proof irrelevance <813UIP>` for the equality type
    defined in the SProp sort, using the :flag:`Definitional UIP` flag,
    (deactivated by default). This models definitional uniqueness of
    identity proofs for this type. It is deactivated by default as
    it can lead to non-termination in combination with impredicativity.

>     Use of this flag is also printed by :cmd:`Print Assumptions`.

Reviewer 2: It's a long explanation, actually longer than the detailed changelog entry. Shouldn't most of this information be moved down to the detailed changelog?

Author: Yes

  Printing of the quotation qualifiers when printing :g:`Ltac` functions
  (`#13028 <https://github.com/coq/coq/pull/13028>`_,
  fixes `#9716 <https://github.com/coq/coq/issues/9716>`_
  and `#13004 <https://github.com/coq/coq/issues/13004>`_,
  by Hugo Herbelin).

SSReflect
^^^^^^^^^

- **Added:**
  SSReflect intro pattern ltac views ``/[dup]``, ``/[swap]`` and ``/[apply]``
  (`#13317 <https://github.com/coq/coq/pull/13317>`_,
  by Cyril Cohen).
- **Fixed:**
  Working around a bug of interaction between + and /(ltac:(...)) cf
  `#13458 <https://github.com/coq/coq/issues/13458>`_
  (`#13459 <https://github.com/coq/coq/pull/13459>`_,
  by Cyril Cohen).

Commands and options
^^^^^^^^^^^^^^^^^^^^

- **Changed:**
  When compiled with OCaml >= 4.10.0, Coq will use the new best-fit GC
  policy, which should provide some performance benefits. Coq's policy
  is optimized for speed, but could increase memory consumption in
  some cases. You are welcome to tune it using the ``OCAMLRUNPARAM``
  variable and report back on good settings so we can improve the defaults.
  (`#13040 <https://github.com/coq/coq/pull/13040>`_,
  fixes `#11277 <https://github.com/coq/coq/issues/11277>`_,

>   by Emilio Jesus Gallego Arias).

Reviewer 2: This entry is misplaced (probably due to the ambiguity of the name "Commands and options"). This entry should be in "Infrastructure and dependencies".

Author: Right.

- **Changed:**
  Drop prefixes from grammar non-terminal names,
  e.g. "constr:global" -> "global", "Prim.name" -> "name".
  Visible in the output of :cmd:`Print Grammar` and :cmd:`Print Custom Grammar`.
  (`#13096 <https://github.com/coq/coq/pull/13096>`_,
  by Jim Fehrle).
- **Changed:**
  When declaring arbitrary terms as hints, unsolved
  evars are not abstracted implicitly anymore and instead
  raise an error
  (`#13139 <https://github.com/coq/coq/pull/13139>`_,
  by Pierre-Marie Pédrot).

> - **Changed:**
>   Option `-native-compiler` of the configure script now impacts the
>   default value of the `-native-compiler` option of coqc. The
>   `-native-compiler` option of the configure script is added as an on demand
>   value, which becomes the default, thus preserving the previous default
>   behavior.
>   The stdlib is still precompiled when configuring with `-native-compiler
>   yes`. It is not precompiled otherwise.
>   This an implementation of point 2 of
>   `CEP #48 <https://github.com/coq/ceps/pull/48>`_
>   (`#13352 <https://github.com/coq/coq/pull/13352>`_,
>   by Pierre Roux).

Reviewer 2: Same problem. This one should probably be in "Tools".

Also, a wording suggestion:
```suggestion
- **Changed:**
  Option `-native-compiler` of the configure script now impacts the
  default value of the `-native-compiler` option of coqc. The
  `-native-compiler` option of the configure script supports a new `ondemand`
  value, which becomes the default, thus preserving the previous default
  behavior.
  The stdlib is still precompiled when configuring with `-native-compiler
  yes`. It is not precompiled otherwise.
  This an implementation of point 2 of
  `CEP #48 <https://github.com/coq/ceps/pull/48>`_
  (`#13352 <https://github.com/coq/coq/pull/13352>`_,
  by Pierre Roux).
```

Author: Done

> - **Changed:**
>   The :attr:`export` locality can now be used for all Hint commands,
>   including :cmd:`Hint Cut`, :cmd:`Hint Mode`, :cmd:`Hint Transparent` / `Opaque` and
>   :cmd:`Remove Hints`
>   (`#13388 <https://github.com/coq/coq/pull/13388>`_,
>   by Pierre-Marie Pédrot).

Reviewer 2: ```suggestion
- **Changed:**
  The :attr:`export` locality can now be used for all Hint commands,
  including :cmd:`Hint Cut`, :cmd:`Hint Mode`, :cmd:`Hint Transparent` / :cmd:`Opaque <Hint Opaque>` and
  :cmd:`Remove Hints`
  (`#13388 <https://github.com/coq/coq/pull/13388>`_,
  by Pierre-Marie Pédrot).
```

Besides, I believe this should actually be an **Added** not a **Changed** (sorry that this also implies moving the location of the changelog).

Author: Right, fixed

- **Removed:**
  In the :cmd:`Extraction Language` command, remove `Ocaml` as a valid value.
  Use `OCaml` instead.  This was deprecated in Coq 8.8, `#6261 <https://github.com/coq/coq/pull/6261>`_
  (`#13016 <https://github.com/coq/coq/pull/13016>`_, by Jim Fehrle).

  .. _813HintWarning:

- **Deprecated:**
  The default value for hint locality is currently :attr:`local` in a section and
  :attr:`global` otherwise, but is scheduled to change in a future release. For the
  time being, adding hints outside of sections without specifying an explicit
  locality is therefore triggering a deprecation warning. It is recommended to
  use :attr:`export` whenever possible
  (`#13384 <https://github.com/coq/coq/pull/13384>`_,
  by Pierre-Marie Pédrot).
- **Deprecated:**
  :cmd:`Grab Existential Variables` and :cmd:`Existential` commands
  (`#12516 <https://github.com/coq/coq/pull/12516>`_,
  by Maxime Dénès).

> - **Added:**
>   Added support for automatic insertion of coercions in :cmd:`Search`

Reviewer 2: ```suggestion
- **Added:**
  Support for automatic insertion of coercions in :cmd:`Search`
```

  patterns. Additionally, head patterns are now automatically
  interpreted as types
  (`#13255 <https://github.com/coq/coq/pull/13255>`_,
  fixes `#13244 <https://github.com/coq/coq/issues/13244>`_,
  by Hugo Herbelin).
- **Added:**
  The :cmd:`Proof using` command can now be used without loading the
  Ltac plugin (`-noinit` mode)
  (`#13339 <https://github.com/coq/coq/pull/13339>`_,
  by Théo Zimmermann).
- **Added:**
  Clarify in the documentation that :cmd:`Add ML Path` is not exported to compiled files
  (`#13345 <https://github.com/coq/coq/pull/13345>`_,
  fixes `#13344 <https://github.com/coq/coq/issues/13344>`_,
  by Hugo Herbelin).

Tools
^^^^^

- **Changed:**
  Added the ability for coq_makefile to directly set the installation folders,

>   through the :n:`COQLIBINSTALL` and :n:`COQDOCINSTALL` variables.

Reviewer 2: ```suggestion
  through the `COQLIBINSTALL` and `COQDOCINSTALL` variables.
```

  See :ref:`coqmakefilelocal`.
  (`#12389 <https://github.com/coq/coq/pull/12389>`_, by Martin Bodin, review of Enrico Tassi).

> - **Changed:**
>   ``dev/tools/make-changelog.sh`` now asks for a list of bugs fixed by the PR
>   (`#12410 <https://github.com/coq/coq/pull/12410>`_, fixes `#12386
>   <https://github.com/coq/coq/issues/12386>`_, by Jason Gross).

Reviewer 2: It's really nice and all, but this is really an internal change of no interest to users reading the changelog, so this should not be there.
```suggestion
```

Author: Indeed, I thought so too.

  m`, :g:`m == n`, :g:`m < n`, :g:`m <= n`, and :g:`m ≤ n` have been replaced
  with :g:`a mod m`, :g:`m =? n`, :g:`m <? n`, :g:`m <=? n`, and :g:`m ≤? n`.
  The old notations are still available as deprecated notations.  Additionally,
  there is now a ``Coq.Numbers.Cyclic.Int63.Int63.Int63Notations`` module that
  users can import to get the ``Int63`` notations without unqualifying the
  various primitives (`#12479 <https://github.com/coq/coq/pull/12479>`_, fixes
  `#12454 <https://github.com/coq/coq/issues/12454>`_, by Jason Gross).
- **Changed:**
  PrimFloat notations now match up with the rest of the standard library: :g:`m
  == n`, :g:`m < n`, and :g:`m <= n` have been replaced with :g:`m =? n`, :g:`m
  <? n`, and :g:`m <=? n`.  The old notations are still available as deprecated
  notations.  Additionally, there is now a
  ``Coq.Floats.PrimFloat.PrimFloatNotations`` module that users can import to
  get the ``PrimFloat`` notations without unqualifying the various primitives
  (`#12556 <https://github.com/coq/coq/pull/12556>`_, fixes `#12454
  <https://github.com/coq/coq/issues/12454>`_, by Jason Gross).

> - **Changed:**
>   Change the sort of cyclic numbers from Type to Set. For backward compatibility, a dynamic sort was defined in the 3 packages bignums, coqprime and color.

Reviewer 2: ```suggestion
- **Changed:** The sort of cyclic numbers from Type to Set.
  For backward compatibility, a dynamic sort was defined in the 3 packages bignums, coqprime and color.
```

  by Olivier Laurent).
- **Added:**
  Extend some list lemmas to both directions: `app_inj_tail_iff`, `app_inv_head_iff`, `app_inv_tail_iff`.
  (`#12094 <https://github.com/coq/coq/pull/12094>`_,
  fixes `#12093 <https://github.com/coq/coq/issues/12093>`_,
  by Edward Wang).
- **Added:**
  ``Decidable`` instance for negation
  (`#12420 <https://github.com/coq/coq/pull/12420>`_,
  by Yishuai Li).
- **Fixed:**
  `Coq.Program.Wf.Fix_F_inv` and `Coq.Program.Wf.Fix_eq` are now axiom-free. They no longer assume proof irrelevance.
  (`#13365 <https://github.com/coq/coq/pull/13365>`_,
  by Li-yao Xia).

Infrastructure and dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **Changed:**
  Coq's core system now uses the `zarith <https://github.com/ocaml/Zarith>`_
  library, based on GNU's gmp instead of ``num`` which is
  deprecated upstream. The custom ``bigint`` module is
  not longer provided; note that ``micromega`` still uses
  ``num``
  (`#11742 <https://github.com/coq/coq/pull/11742>`_,
  by Emilio Jesus Gallego Arias and Vicent Laporte).
- **Removed:**
  The `num` library is not linked to Coq anymore
  (`#13007 <https://github.com/coq/coq/pull/13007>`_,

>   by Emilio Jesus Gallego Arias).

Reviewer 2: The two entries should be merged into a single one, saying that `num` is not a dependency of Coq anymore (not just the core system) and it doesn't need to even mention micromega. BTW, shouldn't Frédéric Besson be credited as well? (e.g., "with help from").

Author: Fixed.

Miscellaneous
^^^^^^^^^^^^^

  .. _813TypingFlags:

- **Added:**
  Typing flags can now be specified per-constant / inductive, this
  allows to fine-grain specify them from plugins or attributes. See
  :ref:`controlling-typing-flags` for details on attribute syntax.
  (`#12586 <https://github.com/coq/coq/pull/12586>`_,

>   by Emilio Jesus Gallego Arias).

Reviewer 2: Should be in "Specification language", not "Misc".

Author: Right.
